gocourse04/main.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"fmt"
+)
+
+const (
+	Ungulates Types = "Ungulates"
+	Birds     Types = "Birds"
+	Primates  Types = "Primates"
+
+	Brush  Purpose = "Grooming Brush"
+	Bucket Purpose = "Water Bucket"
+	Syring Purpose = "Veterinary Syring"
+)
+
+type Zoo struct {
+	areas
+}
+
+func (z Zoo) FindAnimalByName(name string) *Animal {
+	for _, area := range z.areas {
+		for sectorName, sector := range area.sectors {
+			for _, animal := range sector.animals {
+				if animal.name == name {
+					fmt.Printf("Animal %s with id %d is located in area %s, sector %s\n", animal.name, animal.id, area.name, sectorName)
+					return animal
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func NewZoo() *Zoo {
+	temp := make(map[string]Area)
+	return &Zoo{temp}
+}
+
+func (z Zoo) AddArea(t string, area *Area) {
+	z.areas[t] = *area
+}
+
+type Types string
+
+type Area struct {
+	name       string
+	anymalType Types
+	sectors    map[string]*Sector
+}
+
+func NewArea(n string, t Types) *Area {
+	return &Area{n, t, make(map[string]*Sector)}
+}
+
+type areas map[string]Area
+
+type Sector struct {
+	subtype     string
+	animals     []*Animal
+	utilityRoom UtilityRoom
+}
+
+func (s *Sector) AddAnimal(animal *Animal) {
+	s.animals = append(s.animals, animal)
+}
+
+func (s Sector) Describe() {
+	fmt.Printf("Sector subtype is %s.\n", s.subtype)
+	for _, animal := range s.animals {
+		animal.Describe()
+	}
+	s.utilityRoom.Describe()
+}
+
+func (s Sector) FeedAnimal(a Animal) {
+	fmt.Printf("%s sector is feeding the %s animal ....\n", s.subtype, a.name)
+	fmt.Printf("Animal %s is not hungry anymore\n", a.name)
+}
+
+func (a *Area) NewSector(subtype string, u UtilityRoom) *Sector {
+	sector := Sector{subtype: subtype, utilityRoom: u}
+	a.sectors[subtype] = &sector
+	return &sector
+}
+
+type UtilityRoom struct {
+	//The key is a tool's name
+	tools map[string]Tool
+}
+
+func (u UtilityRoom) Describe() {
+	for k, v := range u.tools {
+		fmt.Printf("Tool is %s and purpose is %s\n", k, v.purpose)
+	}
+}
+
+func NewUtilityRoom(t map[string]Tool) *UtilityRoom {
```

Comment:
`tools` - бо багато `Tool`:
```suggestion
func NewUtilityRoom(tools map[string]Tool) *UtilityRoom {
```